core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -0,0 +1,225 @@
+		cache:   map[string]*upkeepStateRecord{},
+		scanner: scanner,
+	}
+}
+
+// Start starts the upkeep state store.
+// it does background cleanup of the cache.
+func (u *upkeepStateStore) Start(pctx context.Context) error {
+	ctx, cancel := context.WithCancel(pctx)
+	defer cancel()
+
+	u.mu.Lock()
+	u.cancel = cancel
+	u.mu.Unlock()
+
+	u.lggr.Debug("Starting upkeep state store")
+
+	ticker := time.NewTicker(GCInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			u.cleanup()
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (u *upkeepStateStore) Close() error {
+	u.mu.Lock()
+	cancel := u.cancel
+	u.mu.Unlock()
+
+	if cancel != nil {
+		cancel()
+	}
+
+	return nil
+}
+
+// SelectByWorkIDs returns the current state of the upkeep for the provided ids.
+// If an id is not found, the state is returned as StateUnknown.
+// We first check the cache, and if any ids are missing, we fetch them from the scanner.
+// TODO: fetch from DB
+func (u *upkeepStateStore) SelectByWorkIDsInRange(ctx context.Context, start, end int64, workIDs ...string) ([]ocr2keepers.UpkeepState, error) {
+	states, ok := u.selectFromCache(workIDs...)
+	if ok {
+		// all ids were found in the cache
+		return states, nil
+	}
+	if err := u.fetchPerformed(ctx, start, end); err != nil {
+		return nil, err
+	}
+	states, _ = u.selectFromCache(workIDs...)
+
+	return states, nil
+}
+
+// SetUpkeepState updates the state of the upkeep.
+// Currently we only store the state if the upkeep is ineligible.
+// Performed events will be fetched on demand.
+func (u *upkeepStateStore) SetUpkeepState(_ context.Context, result ocr2keepers.CheckResult, _ ocr2keepers.UpkeepState) error {
+	if result.Eligible {
+		return nil
+	}
+
+	u.upsertStateRecord(result.WorkID, ocr2keepers.Ineligible, uint64(result.Trigger.BlockNumber))
+
+	return nil
+}
+
+// upsertStateRecord inserts or updates a record for the provided
+// check result. If an item already exists in the data store, the state and
+// block are updated.
+// TODO: persist to DB
+func (u *upkeepStateStore) upsertStateRecord(workID string, s ocr2keepers.UpkeepState, b uint64) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	record, ok := u.cache[workID]
+	if !ok {
+		record = &upkeepStateRecord{
+			workID:  workID,
+			addedAt: time.Now(),
+		}
+	}
+	record.state = s
+	record.block = b
+
+	u.cache[workID] = record
+}
+
+// fetchPerformed fetches all performed logs from the scanner to populate the cache.
+func (u *upkeepStateStore) fetchPerformed(ctx context.Context, start, end int64, workIDs ...string) error {
+	performed, err := u.scanner.WorkIDsInRange(ctx, start, end)
+	if err != nil {
+		return err
+	}
+
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	for _, workID := range performed {
+		if _, ok := u.cache[workID]; !ok {
+			s := &upkeepStateRecord{
+				workID:  workID,
+				state:   ocr2keepers.Performed,
+				addedAt: time.Now(),
+				block:   uint64(end), // TODO: use block number from log
+			}
+			u.cache[workID] = s
```

Comment:
nit. maybe override existing cache value if it was ineligible? Don't think it affects components a lot right now, but perform can take precedence as it is globally accepted truth over local ineligible result.